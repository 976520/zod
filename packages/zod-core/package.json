{
  "name": "zod-core",
  "type": "module",
  "version": "3.23.8",
  "author": "Colin McDonnell <zod@colinhacks.com>",
  "description": "TypeScript-first schema declaration and validation library with static type inference",
  "files": ["src", "dist"],
  "funding": "https://github.com/sponsors/colinhacks",
  "homepage": "https://zod.dev",
  "keywords": ["typescript", "schema", "validation", "type", "inference"],
  "license": "MIT",
  "sideEffects": false,
  "main": "./dist/commonjs/index.js",
  "module": "./dist/esm/index.js",
  "types": "./dist/commonjs/index.d.ts",
  "exports": {
    "./package.json": "./package.json",
    ".": {
      "import": {
        "@zod/source": "./src/index.ts",
        "types": "./dist/esm/index.d.ts",
        "default": "./dist/esm/index.js"
      },
      "require": {
        "types": "./dist/commonjs/index.d.ts",
        "default": "./dist/commonjs/index.js"
      }
    },
    "./util": {
      "import": {
        "@zod/source": "./src/util.ts",
        "types": "./dist/esm/util.d.ts",
        "default": "./dist/esm/util.js"
      },
      "require": {
        "types": "./dist/commonjs/util.d.ts",
        "default": "./dist/commonjs/util.js"
      }
    },
    "./error": {
      "import": {
        "@zod/source": "./src/errors.ts",
        "types": "./dist/esm/errors.d.ts",
        "default": "./dist/esm/errors.js"
      },
      "require": {
        "types": "./dist/commonjs/errors.d.ts",
        "default": "./dist/commonjs/errors.js"
      }
    },
    "./parse": {
      "import": {
        "@zod/source": "./src/parse.ts",
        "types": "./dist/esm/parse.d.ts",
        "default": "./dist/esm/parse.js"
      },
      "require": {
        "types": "./dist/commonjs/parse.d.ts",
        "default": "./dist/commonjs/parse.js"
      }
    },
    "./locales/*": {
      "import": {
        "@zod/source": "./src/locales/*",
        "types": "./dist/esm/locales/*",
        "default": "./dist/esm/locales/*"
      },
      "require": {
        "types": "./dist/commonjs/locales/*",
        "default": "./dist/commonjs/locales/*"
      }
    }
  },
  "tshy": {
    "exports": {
      "./package.json": "./package.json",
      ".": "./src/index.ts",
      "./locales/*": "./src/locales/*"
    },
    "sourceDialects": ["@zod/source"]
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/colinhacks/zod.git"
  },
  "bugs": {
    "url": "https://github.com/colinhacks/zod/issues"
  },
  "support": {
    "backing": {
      "npm-funding": true
    }
  },
  "scripts": {
    "clean": "rm -rf dist",
    "build": "tshy",
    "test:watch": "pnpm vitest",
    "test": "pnpm vitest run",
    "prepublishOnly": "pnpm test && pnpm run build"
  },
  "dependencies": {
    "effect": "^3.5.6"
  }
}
