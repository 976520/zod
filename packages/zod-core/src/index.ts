export * from "./base.js";
export * from "./errors.js";
export * from "./schemas.js";
// export * from "./api.js";
export * from "./regexes.js";
export * from "./checks.js";
