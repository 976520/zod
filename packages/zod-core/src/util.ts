import type { AssertEqual, ErrMessage } from "./types.js";

export function assertEqual<A, B>(val: AssertEqual<A, B>): AssertEqual<A, B> {
  return val;
}
export function assertIs<T>(_arg: T): void {}
export function assertNever(_x: never): never {
  throw new Error();
}

export function arrayToEnum<T extends string, U extends [T, ...T[]]>(
  items: U
): { [k in U[number]]: k } {
  const obj: any = {};
  for (const item of items) {
    obj[item] = item;
  }
  return obj as any;
}

export function getValidEnumValues(obj: any): any {
  const validKeys = objectKeys(obj).filter(
    (k: any) => typeof obj[obj[k]] !== "number"
  );
  const filtered: any = {};
  for (const k of validKeys) {
    filtered[k] = obj[k];
  }
  return objectValues(filtered);
}

export function objectValues(obj: any): any {
  return objectKeys(obj).map((e) => obj[e]);
}

export const objectKeys: ObjectConstructor["keys"] =
  typeof Object.keys === "function"
    ? (obj: any) => Object.keys(obj)
    : (object: any) => {
        const keys = [];
        for (const key in object) {
          if (Object.prototype.hasOwnProperty.call(object, key)) {
            keys.push(key);
          }
        }
        return keys;
      };

export const isInteger: NumberConstructor["isInteger"] =
  typeof Number.isInteger === "function"
    ? (val) => Number.isInteger(val)
    : (val) =>
        typeof val === "number" &&
        Number.isFinite(val) &&
        Math.floor(val) === val;

export function joinValues<T extends any[]>(
  array: T,
  separator = " | "
): string {
  return array
    .map((val) => (typeof val === "string" ? `'${val}'` : val))
    .join(separator);
}
export function jsonStringifyReplacer(_: string, value: any): any {
  if (typeof value === "bigint") {
    return value.toString();
  }
  return value;
}

export function makeCache<This, T extends { [k: string]: () => unknown }>(
  th: This,
  elements: T & ThisType<This>
): { [k in keyof T]: ReturnType<T[k]> } {
  const cache: { [k: string]: unknown } = {};
  for (const key in elements) {
    const getter = elements[key].bind(th);
    Object.defineProperty(cache, key, {
      get() {
        const value = getter();
        Object.defineProperty(cache, key, {
          value,
          configurable: true,
        });
        return value;
      },
      configurable: true,
    });
  }
  return cache as any;
}

export const errToObj = (message?: ErrMessage): { message?: string } =>
  typeof message === "string" ? { message } : message || {};

export const errToString = (message?: ErrMessage): string | undefined =>
  typeof message === "string" ? message : message?.message;
