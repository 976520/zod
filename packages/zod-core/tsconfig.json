{
  "extends": "../../.configs/tsconfig.base.json",
  "compilerOptions": {
    "isolatedDeclarations": true,
    "composite": true,
    "stripInternal": true,
    "strictFunctionTypes": true
  },
  "exclude": ["vitest.config.ts"]
}
