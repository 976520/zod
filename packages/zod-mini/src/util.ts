import type * as core from "zod-core";
import type * as schemas from "./schemas.js";

export type ZodMiniErrorMap<T extends core.$ZodIssueBase = core.$ZodIssueBase> =
  core.$ZodErrorMap<T>;

export type Params<
  T extends schemas.ZodMiniType,
  AlsoOmit extends keyof T["_def"] = never,
> = Partial<Omit<T["_def"], "type" | "checks" | "error" | AlsoOmit>> & {
  error?: string | T["_def"]["error"] | undefined;
};

export interface RawCreateParams {
  error?: string | ZodMiniErrorMap<never> | undefined;
  description?: string | undefined;
  // [k: string]: unknown;
}

export type NormalizedCreateParams<
  T extends RawCreateParams = RawCreateParams,
> = Omit<T, "error"> & {
  error?: ZodMiniErrorMap<core.$ZodIssueBase> | undefined;
  // description?: string | undefined;
};

export function normalizeCreateParams<T extends RawCreateParams>(
  params?: T | undefined
): NormalizedCreateParams<T> {
  const processed: NormalizedCreateParams<T> = {} as any;
  if (!params) return processed;
  const { error: _error, description, ...rest } = params;
  if (_error)
    (processed as any).error =
      typeof _error === "string" ? () => _error : _error;
  if (description) processed.description = description;
  Object.assign(processed, rest);
  return processed;
}

export function splitChecksAndParams<T extends RawCreateParams>(
  _paramsOrChecks?: T | unknown[],
  _checks?: unknown[]
): {
  checks: core.$ZodCheck<any>[];
  params: T;
} {
  const params = (
    Array.isArray(_paramsOrChecks) ? {} : _paramsOrChecks ?? {}
  ) as T;
  const checks: any[] = Array.isArray(_paramsOrChecks)
    ? _paramsOrChecks
    : _checks ?? [];
  return {
    checks,
    params,
  };
}

export interface RawCheckParams {
  error?: string | core.$ZodErrorMap;
}
export interface NormalizedCheckParams {
  error?: core.$ZodErrorMap;
}

export function normalizeCheckParams(
  params?: string | RawCheckParams
): NormalizedCheckParams {
  if (typeof params === "string") return { error: params } as any;
  if (!params) return {} as any;
  return params as any;
}

export interface PrimitiveFactory<
  Params extends RawCreateParams,
  T extends schemas.ZodMiniType,
> {
  (): T;
  (params: Partial<Params>, checks?: core.$ZodCheck<core.output<T>>[]): T;
  (checks: core.$ZodCheck<core.output<T>>[]): T;
}

export const factory: <
  Params extends RawCreateParams,
  T extends schemas.ZodMiniType,
>(
  Cls: any,
  defaultParams: Omit<T["_def"], "checks" | "description" | "error">
) => PrimitiveFactory<Params, T> = (Cls, defaultParams) => {
  return <T extends schemas.ZodMiniType>(...args: any[]) => {
    const { checks, params } = splitChecksAndParams(...args);
    return new Cls({
      ...defaultParams,
      checks,
      ...normalizeCreateParams(params),
    }) as T;
  };
};

function inferNumber<T>(value: T): T {
  return value;
}
const result = inferNumber(0);
