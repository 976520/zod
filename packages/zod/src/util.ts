import type * as core from "zod-core";
import type * as schemas from "./schemas.js";

export type ZodMiniErrorMap<T extends core.$ZodIssueBase = core.$ZodIssueBase> =
  core.$ZodErrorMap<T>;

export type Params<
  T extends schemas.ZodType,
  AlsoOmit extends keyof T["_def"] = never,
> = Partial<Omit<T["_def"], "type" | "checks" | "error" | AlsoOmit>> & {
  error?: string | T["_def"]["error"] | undefined;
};

export interface RawCreateParams {
  error?: string | ZodMiniErrorMap<never> | undefined;
  description?: string | undefined;
  // [k: string]: unknown;
}

export type NormalizedCreateParams<
  T extends RawCreateParams = RawCreateParams,
> = Omit<T, "error"> & {
  error?: ZodMiniErrorMap<core.$ZodIssueBase> | undefined;
  // description?: string | undefined;
};

export function normalizeCreateParams<T extends RawCreateParams>(
  params?: T | undefined
): NormalizedCreateParams<T> {
  const processed: NormalizedCreateParams<T> = {} as any;
  if (!params) return processed;
  const { error: _error, description, ...rest } = params;
  if (_error)
    (processed as any).error =
      typeof _error === "string" ? () => _error : _error;
  if (description) processed.description = description;
  Object.assign(processed, rest);
  return processed;
}

export function splitChecksAndParams<T extends RawCreateParams>(
  _paramsOrChecks?: T | unknown[],
  _checks?: unknown[]
): {
  checks: core.$ZodCheck<any>[];
  params: T;
} {
  const params = (
    Array.isArray(_paramsOrChecks) ? {} : _paramsOrChecks ?? {}
  ) as T;
  const checks: any[] = Array.isArray(_paramsOrChecks)
    ? _paramsOrChecks
    : _checks ?? [];
  return {
    checks,
    params,
  };
}

export interface RawCheckParams {
  error?: string | core.$ZodErrorMap<never>;
  path?: PropertyKey[] | undefined;
}
export interface NormalizedCheckParams {
  error?: core.$ZodErrorMap;
  path?: PropertyKey[] | undefined;
}

export function normalizeCheckParams(
  params?: string | RawCheckParams
): NormalizedCheckParams {
  if (typeof params === "string") return { error: params } as any;
  if (!params) return {} as any;
  if (typeof params.error === "string")
    return { ...params, error: () => params.error } as any;
  return params as any;
}

export interface PrimitiveFactory<
  Params extends RawCreateParams,
  T extends schemas.ZodType,
> {
  (): T;
  (checks: core.$ZodCheck<core.output<T>>[]): T;
  (params: Partial<Params>, checks?: core.$ZodCheck<core.output<T>>[]): T;
}

export const factory: <
  Cls extends { new (...args: any[]): schemas.ZodType },
  Params extends RawCreateParams,
  // T extends InstanceType<Cls> = InstanceType<Cls>,
>(
  Cls: Cls,
  defaultParams: Omit<
    InstanceType<Cls>["_def"],
    "checks" | "description" | "error"
  >
) => PrimitiveFactory<Params, InstanceType<Cls>> = (Cls, defaultParams) => {
  return (...args: any[]) => {
    const { checks, params } = splitChecksAndParams(...args);
    return new Cls({
      ...defaultParams,
      checks,
      ...normalizeCreateParams(params),
    }) as InstanceType<typeof Cls>;
  };
};
