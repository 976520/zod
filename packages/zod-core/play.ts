import * as z from "zod-core";

z;

const schema = z.interface({
  a: z.string(),
  b: z.string(),
  c: z.string(),
  // c: z.number(),
});

const data = {
  a: "asdf",
  b: "asdf",
  c: "asdf",
};
const a = await z.parse(schema, data);
console.log(a);

const b = await z.parseB(schema, data);
console.log(b);
