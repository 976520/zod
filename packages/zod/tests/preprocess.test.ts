// @ts-ignore TS6133
import { expect, test } from "vitest";
import * as util from "zod-core/util";
import * as z from "../src/index.js";

test("preprocess", () => {
  const schema = z.preprocess((data) => [data], z.string().array());

  const value = schema.parse("asdf");
  expect(value).toEqual(["asdf"]);
  util.assertEqual<(typeof schema)["~input"], unknown>(true);
});

test("async preprocess", async () => {
  const schema = z.preprocess(async (data) => {
    return [data];
  }, z.string().array());
  const value = await schema.parseAsync("asdf");
  expect(value).toEqual(["asdf"]);
});

test("preprocess ctx.addIssue with parse", () => {
  const a = z.preprocess((data, ctx) => {
    ctx.addIssue({
      input: data,
      code: "custom",
      message: `${data} is not one of our allowed strings`,
    });
    return data;
  }, z.string());

  const result = a.safeParse("asdf");

  // expect(result.error!.toJSON()).toContain("not one of our allowed strings");
  expect(result.error).toMatchSnapshot();
});

test("preprocess ctx.addIssue non-fatal by default", () => {
  const schema = z.preprocess((data, ctx) => {
    ctx.addIssue({
      code: "custom",
      message: `custom error`,
    });
    return data;
  }, z.string());
  const result = schema.safeParse(1234);
  expect(result.error).toBeInstanceOf(z.ZodError);
  expect(result.error?.issues.length).toEqual(2);
  // expect(result).toMatchSnapshot();
});

test("preprocess ctx.addIssue fatal true", () => {
  const schema = z.preprocess((data, ctx) => {
    ctx.addIssue({
      input: data,
      code: "custom",
      origin: "custom",
      message: `custom error`,
      fatal: true,
    });
    return data;
  }, z.string());

  const result = schema.safeParse(1234);

  expect(result.error).toBeInstanceOf(z.ZodError);
  expect(result.error!.issues.length).toEqual(1);
});

test("async preprocess ctx.addIssue with parseAsync", async () => {
  const schema = z.preprocess(async (data, ctx) => {
    ctx.addIssue({
      input: data,
      code: "custom",
      message: `${data} is not one of our allowed strings`,
    });
    return data;
  }, z.string());

  const result = schema.parseAsync("asdf");
  expect(result).rejects.toThrowErrorMatchingSnapshot();
});

test("z.NEVER in preprocess", () => {
  const foo = z.preprocess((val, ctx) => {
    if (!val) {
      ctx.addIssue({ input: val, code: z.ZodIssueCode.custom, message: "bad" });
      return z.NEVER;
    }
    return val;
  }, z.number());

  type foo = z.infer<typeof foo>;
  util.assertEqual<foo, number>(true);
  const arg = foo.safeParse(undefined);
  if (!arg.success) {
    expect(arg.error.issues[0].message).toEqual("bad");
  }
});

test("preprocess as the second property of object", () => {
  const schema = z.object({
    nonEmptyStr: z.string().min(1),
    positiveNum: z.preprocess((v) => Number(v), z.number().positive()),
  });
  const result = schema.safeParse({
    nonEmptyStr: "",
    positiveNum: "",
  });
  expect(result.success).toEqual(false);
  if (!result.success) {
    expect(result.error.issues.length).toEqual(2);
    expect(result.error.issues[0].code).toEqual(z.ZodIssueCode.too_small);
    expect(result.error.issues[1].code).toEqual(z.ZodIssueCode.too_small);
  }
});

test("preprocess validates with sibling errors", () => {
  const schema = z.object({
    missing: z.string().refine(() => false),
    preprocess: z.preprocess((data: any) => data?.trim(), z.string().regex(/ asdf/)),
  });

  const result = schema.safeParse({ preprocess: " asdf" });
  expect(result.error).toBeInstanceOf(z.ZodError);
  expect(result.error?.issues.length).toEqual(2);
  expect(result.error!.issues).toMatchSnapshot();
});
