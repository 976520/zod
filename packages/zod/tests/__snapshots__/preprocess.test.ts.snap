// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`async preprocess ctx.addIssue with parseAsync 1`] = `
$ZodError {
  "issues": [
    {
      "code": "custom",
      "message": "asdf is not one of our allowed strings",
      "path": [],
    },
  ],
}
`;

exports[`preprocess ctx.addIssue with parse 1`] = `
$ZodError {
  "issues": [
    {
      "code": "custom",
      "message": "asdf is not one of our allowed strings",
      "path": [],
    },
  ],
}
`;

exports[`preprocess validates with sibling errors 1`] = `
[
  {
    "code": "invalid_type",
    "expected": "string",
    "message": "Invalid input: expected string",
    "path": [
      "missing",
    ],
  },
  {
    "code": "invalid_format",
    "format": "regex",
    "message": "Invalid string: must match pattern / asdf/",
    "origin": "string",
    "path": [
      "preprocess",
    ],
    "pattern": "/ asdf/",
  },
]
`;
